### PR TITLE
Fix empty namespace serialization issues

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/1724

### CODE REVIEW

**Comment by joewiz** [comment]
@adamretter Thanks for working on this head scratcher!

I just checked out and built your branch:

```text
project.version=4.2.0-SNAPSHOT
project.built=201806051135
scm.branch=hotfix/empty-namespace-serialization
scm.revision=5ef8b7ac2
```

Unfortunately, when I try the test from my original issue, I am still seeing the same namespace serialization problem. In other words, when I open the resulting `/db/test.xml` via WebDAV or XML-RPC clients (oXygen WebDAV and Java Admin Client), respectively, the document is serialized incorrectly, as:

```xml
<root xmlns="foo">
    <x/>
    <x xml:id="aargh"/>
</root>
```

... instead of as:

```xml
<root xmlns="foo">
    <x/>
    <x xmlns="" xml:id="aargh"/>
</root>
```

... which is how the REST server serializes the document. Specifically, the problem is that `xmlns=""` is omitted from the 2nd `<x>` element when loading this document over WebDAV and XML-RPC.

**Comment by adamretter** [comment]
@joewiz are you sure? I added your example withinor changes to SerialisationTest.java in the PR and that passes now with my changes!

File: src/org/exist/util/serializer/XMLWriter.java [hunk 1]
```diff
@@ -373,47 +373,72 @@ public void comment(final CharSequence data) throws TransformerException {
         }
     }
 
-    public void cdataSection(final char[] ch, final int start, final int len) throws TransformerException {
+    public void startCdataSection() throws TransformerException {
```
Comment (dizzzz) [inline src/org/exist/util/serializer/XMLWriter.java]: should/could be private or package protected when required for unit tests

File: src/org/exist/util/serializer/XMLWriter.java [hunk 2]
```diff
@@ -373,47 +373,72 @@ public void comment(final CharSequence data) throws TransformerException {
...
+
         try {
             writer.write("<![CDATA[");
-            writer.write(ch, start, len);
+        } catch(final IOException ioe) {
+            throw new TransformerException(ioe.getMessage(), ioe);
+        }
+    }
+
+    public void endCdataSection() throws TransformerException {
```
Comment (dizzzz) [inline src/org/exist/util/serializer/XMLWriter.java]: should/could be private or package protected when required for unit tests

File: src/org/exist/storage/NativeBroker.java [hunk 3]
```diff
@@ -2297,7 +2286,7 @@ public Path getBinaryFile(final BinaryDocument blob) throws IOException {
     @Override
     public InputStream getBinaryResource(final BinaryDocument blob)
         throws IOException {
-        return Files.newInputStream(getCollectionFile(getFsDir(), blob.getURI(), false));
+        return Files.newInputStream(getBinaryFile(blob));
```
[the remark this answers is not in the document]
Comment (adamretter) [reply src/org/exist/storage/NativeBroker.java]: Well actually that is not nessecarily the case, `Files.newInputStream` might return a buffered stream. Depends on the OS and JVM. Best to let them decide!

File: src/org/exist/util/serializer/XMLWriter.java (hunk 2 above)
Comment (adamretter) [reply src/org/exist/util/serializer/XMLWriter.java]: Nope. It doesn't override anything.

File: src/org/exist/util/serializer/XMLWriter.java (hunk 1 above)
Comment (adamretter) [reply src/org/exist/util/serializer/XMLWriter.java]: Nope it should be public.

File: src/org/exist/storage/NativeBroker.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dizzzz) [reply src/org/exist/storage/NativeBroker.java]: not according to the java doc ?

**Comment by dizzzz** [comment]
waiting again for the slow AppVeyor

File: src/org/exist/storage/NativeBroker.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamretter) [reply src/org/exist/storage/NativeBroker.java]: @dizzzz Wooow! Okay, I was totally mistaken there. Well there will be many places where this is the case in the codebase. I have not actually changed the buffering here AFAIK.

Perhaps we could cope with this in a future PR from someone... that could look at the uses of `Files.newInputStream` and add buffering where it makes sense for larger transfers?

**Comment by adamretter** [comment]
So @joewiz @dizzzz. I am running the test suite locally, and if Travis-CI passes, I would ask that in this case we make an exception and merge without AppVeyor. There are no Windows specific changes or filesystem changes in this, and I would dearly like to see this go into the 4.2.0 release...

**Pull request merged**